### PR TITLE
Relax stream alias uniqueness

### LINKED ISSUE #219: Relax stream alias uniqueness
[Current](https://github.com/projectriff/system/blob/15fe8359947dcebb288e52e354b2dfaed247860b/pkg/apis/streaming/v1alpha1/processor_validation.go#L118-L148) stream alias validation is across inputs AND outputs, while it could be made across inputs on one side, and across outputs on the other.

### PR DESCRIPTION
**Pull request by AElMehdi**
Fixes #219

### CODE REVIEW

**Comment by fbiville** [comment]
@AElMehdi thanks for the PR! Would you mind referencing #219 in your main commit, as in "Fixes #219"?

File: pkg/apis/streaming/v1alpha1/processor_validation.go [hunk 1]
```diff
@@ -113,12 +113,13 @@ func (s *ProcessorSpec) Validate() validation.FieldErrors {
...
+	errs = errs.Also(s.validateStreamOutputAliasUniqueness())
 
 	return errs
 }
 
-func (s *ProcessorSpec) validateStreamAliasUniqueness() validation.FieldErrors {
+func (s *ProcessorSpec) validateStreamInputAliasUniqueness() validation.FieldErrors {
 	errs := validation.FieldErrors{}
 
 	aliases := []string{}
```
Comment (fbiville) [inline pkg/apis/streaming/v1alpha1/processor_validation.go]: minor: you can write `var aliases []string`

File: pkg/apis/streaming/v1alpha1/processor_validation.go [hunk 2]
```diff
@@ -132,6 +133,21 @@ func (s *ProcessorSpec) validateStreamAliasUniqueness() validation.FieldErrors {
 		}
 		uses[alias] = append(uses[alias], fmt.Sprintf("inputs[%d].alias", i))
 	}
+	for _, alias := range aliases {
+		if len(uses[alias]) > 1 {
+			errs = errs.Also(validation.ErrDuplicateValue(alias, uses[alias]...))
+		}
+	}
+
+	return errs
```
Comment (fbiville) [inline pkg/apis/streaming/v1alpha1/processor_validation.go]: you could still extract common code between Inputs and Outputs, as the validation is the same:
```go
func validateAliasUniqueness(allAliases []string, aliasType string) validation.FieldErrors {
	errs := validation.FieldErrors{}

	var dedupedAliases []string
	uses := map[string][]string{}
	for i, alias := range allAliases {
		if _, ok := uses[alias]; !ok {
			uses[alias] = []string{}
			dedupedAliases = append(dedupedAliases, alias)
		}
		uses[alias] = append(uses[alias], fmt.Sprintf("%s[%d].alias", aliasType, i))
	}

	for _, alias := range dedupedAliases {
		if len(uses[alias]) > 1 {
			errs = errs.Also(validation.ErrDuplicateValue(alias, uses[alias]...))
		}
	}
	return errs
}
```
then, the calling code becomes:
```go
func (s *ProcessorSpec) validateStreamInputAliasUniqueness() validation.FieldErrors {
	var aliases []string
	for _, input := range s.Inputs {
		aliases = append(aliases, input.Alias)
	}
	return validateAliasUniqueness(aliases, "inputs")
}
// [...]
func (s *ProcessorSpec) validateStreamOutputAliasUniqueness() validation.FieldErrors {
	var aliases []string
	for _, output := range s.Outputs {
		aliases = append(aliases, output.Alias)
	}
	return validateAliasUniqueness(aliases, "outputs")
}
```

**Comment by AElMehdi** [comment]
Yeap sure  @fbiville.

**Comment by AElMehdi** [comment]
Let me know if it's ok for you, so I can squash the commits. @fbiville

**Comment by fbiville** [comment]
Merged as 42bc1032637ec93047e5eac05dbca28611f693fd, thanks again @AElMehdi!

**Comment by AElMehdi** [comment]
Sure thing! :relaxed: @fbiville

**Pull request closed without merging**